### PR TITLE
make dev

### PR DESCRIPTION
**Pull request by alsuren**
After my experience on the FutureNHS project, I've come to believe that the local development experience is pretty crucial.

It's also the middle of summer, so I've been working in the only cool room in the house, with my laptop on my knee. I've therefore been reluctant to run kubernetes locally https://github.com/kubernetes/kubernetes/issues/100415 .

I decided to try out a fully local development setup, without any docker in the loop.

`dapr init` installs redis, zipkin and something called the "placement service" as docker containers by default. If you aren't in the mood for running docker, you can `dapr init --slim` and it works fine without (you will need to nuke your ~/.dapr/components first though).

~If anyone knows of a tool that's less annoying than GNU parallel, please shout.~ `pnpx concurrently` it is

- [x] point to a local dapr config, with spec: tracing: samplingRate: "0"

### CODE REVIEW

**Comment by ceddlyburge** [comment]
Hi David, more efficient code is better for the environment as well :)

I did the following

- brew install parrallel
- delete .dapr/components
- dapr init --slim
- make dev

The hr service span up ok, but the `make dev` command repeatedly prints out this error (and I can't connect to the slack service any other way either)

> 2021/06/15 12:28:01 failed to send the request: Post "http://localhost:9411/api/v2/spans": dial tcp [::1]:9411: connect: connection refused

Am I missing something?

And could we update the readme?

Thanks, Cedd

**Comment by alsuren** [comment]
> 2021/06/15 12:28:01 failed to send the request: Post "http://localhost:9411/api/v2/spans": dial tcp [::1]:9411: connect: connection refused

that is it trying to connect to zipkin... ah yes. I remember now. I had to update my ~/.dapr/config.yml to include
```
spec:
  tracing:
    samplingRate: "0"
```

- [x] I will have to have a think about this. Relying on changes to configs that live outside the repo makes me super-sad. I'll check if there's a better way to do this and get back to you.

**Comment by ceddlyburge** [comment]
That does stop the errors for me, and I can now hit the slack api, which can in turn talk to the hr api.

`curl http://localhost:3001/ping`

But when I try to invoke it via dapr I get "{"errorCode":"ERR_DIRECT_INVOKE","message":"fail to invoke, id: slack, err: couldn't find service: slack"}"

`dapr invoke --app-id hr --method ping`

Invoking the curl gives the same problem

curl -X POST -H "Content-Type: application/x-www-form-urlencoded" --data @lib/slack/exampleRequest.txt  http://localhost:62131/v1.0/invoke/slack.slack/method/getManager

**Comment by alsuren** [comment]
> But when I try to invoke it via dapr I get "{"errorCode":"ERR_DIRECT_INVOKE","message":"fail to invoke, id: slack, err: couldn't find service: slack"}"

If you try to run the slack service (`make -C lib/slack/ dev`), do you get `❌  dapr with ID slack is already running`, or do you get some other error?

I have been using yarn 2 locally, so I might have missed something here. Maybe I will switch us to yarn 2 and use `concurrently` rather than `parallel`, and update the lockfile etc. Objections?

**Comment by ceddlyburge** [comment]
> If you try to run the slack service (make -C lib/slack/ dev), do you get ❌ dapr with ID slack is already running, or do you get some other error?

I get the ❌ dapr with ID slack is already running

no objections to yarn 2 ...

**Comment by ceddlyburge** [comment]
Hi David, is there anything for me to do on this PR? I think this is the first of the PRs that you want to merge. I could check it out and go through the readme and test it again?

File: lib/slack/Dockerfile [hunk 1]
```diff
@@ -1,19 +1,17 @@
...
 
-FROM node:lts-alpine AS builder
-
-WORKDIR /opt/app
-ENV NODE_ENV=production
 COPY . .
-COPY --from=deps /opt/app/node_modules ./node_modules
-RUN yarn build
+# For pnpm, this is just hardlinking a bunch of directories, so it's basically free.
+RUN pnpm install --offline --dev
```
Comment (alsuren) [inline lib/slack/Dockerfile]:
```suggestion
RUN pnpm install --offline
```

backporting this from the other PR

**Comment by alsuren** [comment]
> Hi David, is there anything for me to do on this PR? I think this is the first of the PRs that you want to merge. I could check it out and go through the readme and test it again?

You said that you had some problems getting it working before. I *think* that the switch from `parallel` to `concurrently` should make things a bit better. If you want to test it out again, you can. Otherwise, I have been working on top of this branch for a while, so I've probably ironed out all of the big problems. If you think that's enough, you can just approve it and I'll merge.

**Comment by ceddlyburge** [comment]
I ran through the "Running on local Dapr Instance" and it all worked fine.

**Review by ceddlyburge — APPROVED** [review]

**Pull request merged**